profile/encode.go

```diff
@@ -231,34 +230,54 @@ var profileDecoder = []decoder{
 // The unexported fields are cleared up to facilitate testing.
 func (p *Profile) postDecode() error {
 	var err error
-	mappings := make(map[uint64]*Mapping)
+	mappings := make(map[uint64]*Mapping, len(p.Mapping))
+	mappingIds := make([]*Mapping, len(p.Mapping)+1)
 	for _, m := range p.Mapping {
 		m.File, err = getString(p.stringTable, &m.fileX, err)
 		m.BuildID, err = getString(p.stringTable, &m.buildIDX, err)
+		if m.ID < uint64(len(mappingIds)) {
+			mappingIds[m.ID] = m
+			continue
+		}
 		mappings[m.ID] = m
 	}
 
-	functions := make(map[uint64]*Function)
+	functions := make(map[uint64]*Function, len(p.Function))
+	functionIds := make([]*Function, len(p.Function)+1)
 	for _, f := range p.Function {
 		f.Name, err = getString(p.stringTable, &f.nameX, err)
 		f.SystemName, err = getString(p.stringTable, &f.systemNameX, err)
 		f.Filename, err = getString(p.stringTable, &f.filenameX, err)
+		if f.ID < uint64(len(functionIds)) {
+			functionIds[f.ID] = f
+			continue
```

Comment:
in this case, it seems to me that explicit if - else is both more readable and safer as it is easier to add code after the blocks or move blocks without reevaluating the loop's short-circuit logic (as demonstrated in lines 261-265).